### DISCUSSION TITLE
What's the best way to performantly have parallelized calls into a WASM module?

### DISCUSSION THREAD

**Discussion by john01dav** [wasmerio/wasmer#4382] [General]
I am working on a high performance program which needs to load plugins. I would like to use Rust for plugins, and load them via WASM. But, I need to be able to saturate all CPU cores with the WASM code in the same module (the problem is almost embarrassingly parallel). This poses a problem since each WASM function call requires a &mut Store, so I can't use the same Store across threads. I could create a separate Instance and Store for each thread, but then the static state in each plugin is not consistent which is of course not ideal for the programmers of those plugins. 

Is there some established way of doing this without the aforementioned problems?

**Comment by linear** [comment]
<p><a href="https://linear.app/wasmer/issue/RUN-31/whats-the-best-way-to-performantly-have-parallelized-calls-into-a-wasm">RUN-31 What's the best way to performantly have parallelized calls into a WASM module?</a></p>

**Comment by Arshia001** [comment]
@john01dav an `Instance` is inherently single-threaded. Your options here are to either:
* run a separate instance per thread, and make sure each instance is getting enough work from the native code, or
* use WASIX threads inside the plugins. This requires the plugins to be compiled to WASIX.
